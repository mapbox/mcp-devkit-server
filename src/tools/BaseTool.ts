import {
  McpServer,
  RegisteredTool
} from '@modelcontextprotocol/sdk/server/mcp.js';
import { RequestHandlerExtra } from '@modelcontextprotocol/sdk/shared/protocol.js';
import { ToolAnnotations } from '@modelcontextprotocol/sdk/types.js';
import { z, ZodTypeAny } from 'zod';

const ContentItemSchema = z.union([
  z.object({
    type: z.literal('text'),
    text: z.string()
  }),
  z.object({
    type: z.literal('image'),
    data: z.string(),
    mimeType: z.string()
  })
]);

export const OutputSchema = z.object({
  content: z.array(ContentItemSchema),
  isError: z.boolean().default(false)
});

export type ContentItem = z.infer<typeof ContentItemSchema>;
export type ToolOutput = z.infer<typeof OutputSchema>;

export abstract class BaseTool<InputSchema extends ZodTypeAny> {
  abstract readonly name: string;
  abstract readonly description: string;
  abstract readonly annotations: ToolAnnotations;

  readonly inputSchema: InputSchema;
  protected server: McpServer | null = null;

  constructor(params: { inputSchema: InputSchema }) {
    this.inputSchema = params.inputSchema;
  }

  /**
   * Validates and runs the tool logic.
   */
  async run(
    rawInput: unknown,
    extra?: RequestHandlerExtra<any, any>
  ): Promise<z.infer<typeof OutputSchema>> {
    try {
      const input = this.inputSchema.parse(rawInput);
      const accessToken =
        extra?.authInfo?.token || process.env.MAPBOX_ACCESS_TOKEN;
      const result = await this.execute(input, accessToken);

      // Check if result is already a content object (image or text)
      if (
        result &&
        typeof result === 'object' &&
        'type' in result &&
        (result.type === 'image' || result.type === 'text')
      ) {
        return {
          content: [result as ContentItem],
          isError: false
        };
      }

      // Otherwise return as text
      return {
        content: [{ type: 'text', text: JSON.stringify(result) }],
        isError: false
      };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);

      this.log(
        'error',
        `${this.name}: Error during execution: ${errorMessage}`
      );

      return {
        content: [
          {
            type: 'text',
            text: errorMessage || 'Internal error has occurred.'
          }
        ],
        isError: true
      };
    }
  }

  /**
   * Tool logic to be implemented by subclasses.
   */
  protected abstract execute(
    _input: z.infer<InputSchema>,
    accessToken?: string
  ): Promise<ContentItem | unknown>;

  /**
   * Installs the tool to the given MCP server.
   */
  installTo(server: McpServer): RegisteredTool {
    this.server = server;

    return server.registerTool(
      this.name,
      {
        description: this.description,
        inputSchema: (
          this.inputSchema as unknown as z.ZodObject<
            Record<string, z.ZodTypeAny>
          >
        ).shape,
        annotations: this.annotations
      },
      (args: any, extra: any) => this.run(args, extra)
    );
  }

  /**
   * Helper method to send logging messages
   */
  protected log(
    level: 'debug' | 'info' | 'warning' | 'error',
    data: string | Record<string, unknown>
  ): void {
    if (this.server) {
      this.server.server.sendLoggingMessage({ level, data });
    }
  }
}
