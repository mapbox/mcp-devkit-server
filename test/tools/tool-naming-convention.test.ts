import { describe, it, expect } from 'vitest';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { BaseTool } from '../../src/tools/BaseTool.js';
import { pathToFileURL } from 'node:url';

async function discoverTools(): Promise<any[]> {
  const toolsDir = path.resolve(
    new URL('.', import.meta.url).pathname,
    '../../src/tools'
  );
  const tools: any[] = [];

  // Find all directories that end with '-tool'
  const entries = fs.readdirSync(toolsDir, { withFileTypes: true });
  const toolDirectories = entries
    .filter((entry) => entry.isDirectory() && entry.name.endsWith('-tool'))
    .map((entry) => entry.name);

  // Dynamically import each tool
  for (const toolDir of toolDirectories) {
    const allFiles = fs.readdirSync(path.join(toolsDir, toolDir));
    const toolFiles = allFiles.filter(
      (file) => file.endsWith('.ts') && file.includes('Tool')
    );

    // Import all tool files in the directory
    for (const toolFile of toolFiles) {
      try {
        const modulePath = pathToFileURL(
          path.join(toolsDir, toolDir, toolFile.replace('.ts', '.js'))
        ).href;
        const module = await import(modulePath);

        // Find all exported tool classes
        const toolClasses = Object.values(module).filter(
          (exp: any) =>
            exp && exp.prototype && exp.prototype instanceof BaseTool
        );

        for (const toolClass of toolClasses) {
          tools.push(new (toolClass as any)());
        }
      } catch (error) {
        console.warn(
          `Failed to import tool from ${toolDir}/${toolFile}: ${error}`
        );
      }
    }
  }

  return tools;
}

describe('Tool Naming Convention', () => {
  it('should enforce snake_case naming with _tool suffix for all tool names', async () => {
    // Snake case pattern: lowercase letters, numbers, and underscores only
    // Must start with a letter, no consecutive underscores
    // Must end with "_tool" suffix
    const snakeCasePattern = /^[a-z][a-z0-9]*(_[a-z0-9]+)*$/;
    const toolSuffixPattern = /_tool$/;

    // Dynamically discover all tool instances
    const tools = await discoverTools();

    const violations: string[] = [];

    expect(tools.length).toBeGreaterThan(0); // Ensure we found some tools

    // Check each tool's name
    for (const tool of tools) {
      const toolName = tool.name;
      const className = tool.constructor.name;

      if (!snakeCasePattern.test(toolName)) {
        violations.push(
          `${className}: "${toolName}" is not in snake_case format`
        );
      }

      if (!toolSuffixPattern.test(toolName)) {
        violations.push(
          `${className}: "${toolName}" must end with "_tool" suffix`
        );
      }
    }

    // Report all violations
    if (violations.length > 0) {
      const errorMessage = [
        'Tool names must follow snake_case convention and end with "_tool":',
        '- Use lowercase letters and numbers only',
        '- Separate words with single underscores',
        '- Start with a letter',
        '- No consecutive underscores',
        '- Must end with "_tool" suffix',
        '',
        `Found ${tools.length} tools to check.`,
        '',
        'Violations found:',
        ...violations.map((v) => `  - ${v}`),
        '',
        'Please update the tool name property to follow the naming convention.'
      ].join('\n');

      throw new Error(errorMessage);
    }
  });

  it('should maintain consistent tool list (snapshot test)', async () => {
    // Dynamically discover all tool instances
    const tools = await discoverTools();

    // Extract tool names and class names for snapshot
    const toolSnapshot = tools
      .map((tool) => ({
        className: tool.constructor.name,
        toolName: tool.name,
        description: tool.description
      }))
      .sort((a, b) => a.className.localeCompare(b.className));

    // This snapshot test will fail if tools are accidentally added or removed
    expect(toolSnapshot).toMatchSnapshot();

    // Also verify we have the expected number of tools
    expect(tools.length).toBeGreaterThan(0);
  });

  it('should validate snake_case pattern with _tool suffix examples', () => {
    const snakeCasePattern = /^[a-z][a-z0-9]*(_[a-z0-9]+)*$/;
    const toolSuffixPattern = /_tool$/;

    // Valid tool names (snake_case + _tool suffix)
    const validToolNames = [
      'my_tool',
      'create_style_tool',
      'list_tokens_tool',
      'bounding_box_tool',
      'simple_tool',
      'coordinate_conversion_tool'
    ];

    // Invalid tool names (violate snake_case or missing _tool suffix)
    const invalidToolNames = [
      'tool', // missing _tool suffix
      'list_tokens', // missing _tool suffix
      'bounding_box', // missing _tool suffix
      'MyTool', // not snake_case
      'my-tool', // uses hyphens
      'my__tool', // consecutive underscores
      'my_tool_', // trailing underscore after _tool
      '_my_tool', // leading underscore
      'my tool', // contains space
      'TOOL', // all uppercase
      'myTool' // camelCase
    ];

    // Test valid tool names (must pass both patterns)
    for (const name of validToolNames) {
      expect(snakeCasePattern.test(name)).toBe(true);
      expect(toolSuffixPattern.test(name)).toBe(true);
    }

    // Test invalid tool names (should fail at least one pattern)
    for (const name of invalidToolNames) {
      const isValidSnakeCase = snakeCasePattern.test(name);
      const hasToolSuffix = toolSuffixPattern.test(name);
      expect(isValidSnakeCase && hasToolSuffix).toBe(false);
    }
  });
});
