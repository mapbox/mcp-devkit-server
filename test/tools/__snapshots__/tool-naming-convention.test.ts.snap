// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tool Naming Convention > should maintain consistent tool list (snapshot test) 1`] = `
[
  {
    "className": "BoundingBoxTool",
    "description": "Calculates bounding box of given GeoJSON content, returns as [minX, minY, maxX, maxY]",
    "toolName": "bounding_box_tool",
  },
  {
    "className": "CoordinateConversionTool",
    "description": "Converts coordinates between WGS84 (longitude/latitude) and EPSG:3857 (Web Mercator) coordinate systems",
    "toolName": "coordinate_conversion_tool",
  },
  {
    "className": "CountryBoundingBoxTool",
    "description": "Gets bounding box for a country by its ISO 3166-1 country code, returns as [minX, minY, maxX, maxY].",
    "toolName": "country_bounding_box_tool",
  },
  {
    "className": "CreateStyleTool",
    "description": "Create a new Mapbox style",
    "toolName": "create_style_tool",
  },
  {
    "className": "CreateTokenTool",
    "description": "Create a new Mapbox public access token with specified scopes and optional URL restrictions.",
    "toolName": "create_token_tool",
  },
  {
    "className": "DeleteStyleTool",
    "description": "Delete a Mapbox style by ID",
    "toolName": "delete_style_tool",
  },
  {
    "className": "GeojsonPreviewTool",
    "description": "Generate a geojson.io URL to visualize GeoJSON data. Returns only the URL link.",
    "toolName": "geojson_preview_tool",
  },
  {
    "className": "GetMapboxDocSourceTool",
    "description": "Get the latest official Mapbox documentation, APIs, SDKs, and developer resources directly from Mapbox. Always up-to-date, comprehensive coverage of all current Mapbox services including mapping, navigation, search, geocoding, and mobile SDKs. Use this for accurate, official Mapbox information instead of web search.",
    "toolName": "get_latest_mapbox_docs_tool",
  },
  {
    "className": "ListStylesTool",
    "description": "List styles for a Mapbox account. Use limit parameter to avoid large responses (recommended: limit=5-10). Use start parameter for pagination.",
    "toolName": "list_styles_tool",
  },
  {
    "className": "ListTokensTool",
    "description": "List Mapbox access tokens for the authenticated user with optional filtering and pagination. When using pagination, the "start" parameter must be obtained from the "next_start" field of the previous response (it is not a token ID)",
    "toolName": "list_tokens_tool",
  },
  {
    "className": "PreviewStyleTool",
    "description": "Generate preview URL for a Mapbox style using an existing public token",
    "toolName": "preview_style_tool",
  },
  {
    "className": "RetrieveStyleTool",
    "description": "Retrieve a specific Mapbox style by ID",
    "toolName": "retrieve_style_tool",
  },
  {
    "className": "StyleBuilderTool",
    "description": "Generate Mapbox style JSON for creating new styles or updating existing ones.

The tool intelligently resolves layer types and filter properties using Streets v8 data.
You don't need exact layer names - the tool automatically finds the correct layer based on your filters.

BASE STYLES:
• standard: ALWAYS THE DEFAULT - Modern Mapbox Standard with best performance
• Classic styles: streets-v11/light-v10/dark-v10/satellite-v9/outdoors-v11
  Only use Classic when user explicitly says "create a classic style" or working with existing Classic style

STANDARD STYLE CONFIG:
Use standard_config to customize the basemap:
• Theme: default/faded/monochrome
• Light: day/night/dawn/dusk
• Show/hide: labels, roads, 3D buildings
• Colors: water, roads, parks, etc.

LAYER ORDERING:
• Layers are rendered in order - last layer in the array appears visually on top
• The 'slot' parameter is OPTIONAL - by default, layer order in the array determines visibility
• For Standard style, you can optionally use 'slot' to control placement:
  - No slot (default): Above all existing layers in the style
  - 'top': Behind Place and Transit labels
  - 'middle': Between basemap and labels
  - 'bottom': Below most basemap features

LAYER RENDERING:
• render_type controls HOW to visualize the layer (line, fill, symbol, etc.)
• Most important: Use render_type:"line" for outlines/borders even on polygon features
• Default "auto" picks based on geometry, but override for specific effects:
  - Building outlines → render_type:"line" (not fill!)
  - Solid buildings → render_type:"fill" or "fill-extrusion" (3D)
  - Road lines → render_type:"line" (auto works too)
  - POI dots → render_type:"circle"
  - Labels → render_type:"symbol"

LAYER ACTIONS:
• color: Apply a specific color
• highlight: Make prominent
• hide: Remove from view
• show: Display with defaults

AUTO-DETECTION:
The tool automatically finds the correct layer from your filter_properties.
Examples:
• { class: 'park' } → finds 'landuse' layer
• { type: 'wetland' } → finds 'landuse_overlay' layer
• { maki: 'restaurant' } → finds 'poi_label' layer
• { toll: true } → finds 'road' layer
• { admin_level: 0 } → finds 'admin' layer (for country boundaries)
• { admin_level: 1 } → finds 'admin' layer (for state/province boundaries)

IMPORTANT LAYER NAMES:
• Use "admin" for all boundaries (countries, states, etc.)
• Use "building" (singular, not "buildings")
• Use "road" for all streets, highways, paths

If a layer type is not recognized, the tool will provide helpful suggestions showing:
• All available source layers from Streets v8
• Which fields are available in each layer
• Examples of how to properly specify layers and filters",
    "toolName": "style_builder_tool",
  },
  {
    "className": "StyleComparisonTool",
    "description": "Generate a comparison URL for comparing two Mapbox styles side-by-side",
    "toolName": "style_comparison_tool",
  },
  {
    "className": "TilequeryTool",
    "description": "Query vector and raster data from Mapbox tilesets at geographic coordinates",
    "toolName": "tilequery_tool",
  },
  {
    "className": "UpdateStyleTool",
    "description": "Update an existing Mapbox style",
    "toolName": "update_style_tool",
  },
]
`;
